pandas/core/indexes/period.py

```diff
@@ -148,144 +177,176 @@ class PeriodIndex(PeriodArrayMixin, DatelikeOps, DatetimeIndexOpsMixin,
-            raise TypeError("PeriodIndex does not allow "
-                            "floating point in construction")
-
-        # anything else, likely an array of strings or periods
-        data = ensure_object(data)
-        freq = freq or period.extract_freq(data)
-        data = period.extract_ordinals(data, freq)
-        return cls._from_ordinals(data, name=name, freq=freq)
+    @property
+    def values(self):
+        return self._data
 
-    @cache_readonly
-    def _engine(self):
-        return self._engine_type(lambda: self, len(self))
+    @property
+    def freq(self):
+        # TODO(DatetimeArray): remove
+        # Can't simply use delegate_names since our base class is defining
+        # freq
+        return self._data.freq
+
+    @freq.setter
+    def freq(self, value):
+        value = Period._maybe_convert_freq(value)
+        msg = ('Setting {cls}.freq has been deprecated and will be '
+               'removed in a future version; use {cls}.asfreq instead. '
+               'The {cls}.freq setter is not guaranteed to work.')
+        warnings.warn(msg.format(cls=type(self).__name__),
+                      FutureWarning, stacklevel=2)
+        # PeriodArray._freq isn't actually mutable. We set the private _freq
+        # here, but people shouldn't be doing this anyway.
+        self._data._freq = value
+
+    # ------------------------------------------------------------------------
+    # Index Constructors
 
     @classmethod
     def _simple_new(cls, values, name=None, freq=None, **kwargs):
+        # type: (PeriodArray, Any, Any) -> PeriodIndex
         """
         Values can be any type that can be coerced to Periods.
         Ordinals in an ndarray are fastpath-ed to `_from_ordinals`
         """
-        if not is_integer_dtype(values):
-            values = np.array(values, copy=False)
-            if len(values) > 0 and is_float_dtype(values):
-                raise TypeError("PeriodIndex can't take floats")
-            return cls(values, name=name, freq=freq, **kwargs)
-
-        return cls._from_ordinals(values, name, freq, **kwargs)
+        assert isinstance(values, PeriodArray)
+        result = object.__new__(cls)
+        result._data = values
+        result.name = name
+        result._reset_identity()
+        return result
 
     @classmethod
     def _from_ordinals(cls, values, name=None, freq=None, **kwargs):
         """
         Values should be int ordinals
         `__new__` & `_simple_new` cooerce to ordinals and call this method
         """
-        result = super(PeriodIndex, cls)._from_ordinals(values, freq)
-
-        result.name = name
-        result._reset_identity()
+        data = PeriodArray._from_ordinals(values, freq=freq)
+        result = cls._simple_new(data, name=name)
         return result
 
-    def _shallow_copy_with_infer(self, values, **kwargs):
+    def _shallow_copy(self, values=None, **kwargs):
+        # TODO: simplify, figure out type of values
```

Comment:
So from some printing in the tests, some exploration on what is passed here:

- PeriodArray and int64 ndarray ordinals. I think those both are fine, it will probably hard to avoid mixing both? Or do we want a separate one for ordinals?
- object array of Periods. 
  - One example of this is `PeriodIndex.difference` (from the base Index implementation). This base implementation basically works, except that there is a `sorting.safe_sort` call on the resulting PeriodArray, which destroys the PeriodArray. But this is of course solvable in `sorting.safe_sort`, by making that EA aware.
  - So I think eventually we could try to solve all those cases where object is passed. But I would say, let's leave that for follow-ups ?
- None -> this is from plain `self.shallow_copy()` calls without arguments. This is fine I think.

Reply:
This is basically what motivated #23095.  Even if the solution is unwanted there, I think it identifies all the extant places where unwanted types are currently passed to _shallow_copy